app/src/swap.c

```diff
@@ -0,0 +1,249 @@
+
+    if (strncmp(tmp_str, &valid_tx_method[0], strlen(valid_tx_method)) != 0) {
+         PRINTF("Wrong swap tx method (%s, should be : %s).\n",tmp_str,valid_tx_method);
+         ZEMU_LOGF(100, "Wrong swap tx method (%s, should be : %s).\n",tmp_str,valid_tx_method)
+         return parser_swap_tx_wrong_method;
+    }
+
+    // Check transaction method arguments number. Balance transfer Should be 2 (for tx v18).
+    if (_getMethod_NumItems(ctx->tx_obj->transactionVersion,
+                           ctx->tx_obj->callIndex.moduleIdx,
+                           ctx->tx_obj->callIndex.idx) != 2) {
+        ZEMU_LOGF(50, "Wrong swap tx method arguments count.\n")
+        return parser_swap_tx_wrong_method_args_num;
+    }
+
+    // Check destination address.
+    MEMZERO(tmp_str,sizeof(tmp_str));
+    uint8_t pageCount = 0;
+    err = _getMethod_ItemValue(ctx->tx_obj->transactionVersion,
+                            &ctx->tx_obj->method,
+                            ctx->tx_obj->callIndex.moduleIdx, ctx->tx_obj->callIndex.idx, 0,
+                            tmp_str, sizeof(tmp_str),
+                            0, &pageCount);
+
+    if (err != parser_ok) {
+        ZEMU_LOGF(50, "Could not parse swap tx destination address.");
+        return err;
+    }
+
+    if (strncmp(tmp_str, &(G_swap_state.destination_address[0]), sizeof(G_swap_state.destination_address)) != 0) {
+         ZEMU_LOGF(100, "Wrong swap tx destination address (%s, should be : %s).\n", tmp_str, G_swap_state.destination_address)
+         return parser_swap_tx_wrong_dest_addr;
+    }
+
+    // Check amount.
+    MEMZERO(tmp_str, sizeof(tmp_str));
+    err = _getMethod_ItemValue(ctx->tx_obj->transactionVersion,
+                            &ctx->tx_obj->method,
+                            ctx->tx_obj->callIndex.moduleIdx, ctx->tx_obj->callIndex.idx, 1,
+                            tmp_str, sizeof(tmp_str),
+                            0, &pageCount);
+
+    if(err != parser_ok)
+    {
+        ZEMU_LOGF(50, "Could not parse swap tx amount.");
+        return err;
+    }
+
+    char tmp_amount[50] = {0};
+    pd_CompactBalance_t amount = {.value.ptr = G_swap_state.amount, .value.len = G_swap_state.amount_length};
+    err = _toStringCompactBalance(&amount, tmp_amount, sizeof(tmp_amount), 0, &pageCount);
+
+    if (err != parser_ok || strncmp(tmp_str, tmp_amount, sizeof(tmp_amount)) != 0) {
+        ZEMU_LOGF(100, "Wrong swap tx amount (%s, should be : %s).\n", tmp_str, tmp_amount)
+        return parser_swap_tx_wrong_amount;
+    }
+
+    ZEMU_LOGF(50, "Swap parameters verified by current tx\n")
+    return err;
+}
+
+
+////////////////////////////////////////////////////////////////
+zxerr_t extractHDPath(uint8_t* params, uint8_t paramsSize) {
+    if (paramsSize != (sizeof(uint32_t) * HDPATH_LEN_DEFAULT)) {
+        return zxerr_invalid_crypto_settings;
+    }
+
+    for (uint32_t i = 0; i < paramsSize; i+= 4) {
+        if (readU32BE(params + i, &hdPath[i]) != zxerr_ok) {
```

Comment:
You should not use `i` (which is incremented by 4) as your `hdPath` index but rather another variable that you would increment by one. @ftheirs @carlosala

Reply:
Fixed in b7ed534d533292268c3d1010a0e3c3efc7445594